sbi/neural_nets/embedding_nets.py

```diff
@@ -266,19 +274,36 @@ def forward(self, x: Tensor) -> Tensor:
         Returns:
             Network output (batch_size, output_dim).
         """
-        batch, permutation_dim, _ = x.shape
 
-        iid_embeddings = self.trial_net(x.view(batch * permutation_dim, -1)).view(
-            batch, permutation_dim, -1
+        # Get number of trials from non-nan entries
+        num_batch, max_num_trials = x.shape[0], x.shape[self.aggregation_dim]
+        nan_counts = (
+            torch.isnan(x)
+            .sum(dim=self.aggregation_dim)  # count nans over trial dimension
+            .reshape(-1)[:num_batch]  # counts are the same across data dims
+            .unsqueeze(-1)  # make it (batch, 1) to match embeddings below
         )
-
-        if self.combining_operation == "mean":
-            e = iid_embeddings.mean(dim=1)
-        elif self.combining_operation == "sum":
-            e = iid_embeddings.sum(dim=1)
+        # number of non-nan trials
+        trial_counts = max_num_trials - nan_counts
+
+        # get nan entries
+        is_nan = torch.isnan(x)
+        # apply trial net with nan entries replaced with 0
+        masked_x = torch.nan_to_num(x, nan=0.0)
+        trial_embeddings = self.trial_net(masked_x)
+        # replace previous nan entries with zeros
+        trial_embeddings = trial_embeddings * (~is_nan.all(-1, keepdim=True)).float()
+
+        # Take mean over permutation dimension divide by number of trials
+        # (instead of just taking torch.mean) to account for masking.
+        if self.aggregation_fn == "mean":
+            combined_embedding = (
+                trial_embeddings.sum(dim=self.aggregation_dim) / trial_counts
+            )
         else:
-            raise ValueError("combining_operation must be in ['sum', 'mean'].")
+            combined_embedding = trial_embeddings.sum(dim=self.aggregation_dim)
 
-        embedding = self.fc_subnet(e)
+        assert not torch.isnan(combined_embedding).any(), "NaNs in embedding."
```

Comment:
this will ensure that even if `exclude_invalid_x = False`, if there are NaNs somewhere else in the data (other than for masking trials), training will not proceed.